compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,199 @@ static bool consumesInputOp(NamedSequenceOp op) {
+                                 tuningSpecs.namedSequenceToUser);
+    }
+  }
+
+  return tuningSpecs;
+}
+
+// Renames a `NamedSequenceOp` to resolve name conflicts caused by merging
+// tuning specs, moves it to the top-level module, and updates its occurrences
+// in corresponding `ForeachMatchOp` operations.
+static void updateNamedSequenceOp(
+    NamedSequenceOp op, ModuleOp module, OpBuilder &builder,
+    llvm::StringMap<unsigned> &specNameCounts,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser) {
+  StringRef specName = op.getSymName();
+  std::string newSpecName = getUniqueSpecName(specName, specNameCounts);
+  op.setSymName(newSpecName);
+
+  // Skip updating ForeachMatchOp if the NamedSequenceOp is not used in it
+  // or its name has not changed.
+  if (!namedSequenceToUser.contains(op) || newSpecName == specName) {
+    op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+    return;
+  }
+
+  ForeachMatchOp foreachMatchOp = namedSequenceToUser[op];
+
+  // Helper function for updating matchers or actions in associated
+  // `foreachMatchOp` instances.
+  auto getUpdatedSymbol = [&](Attribute attr) -> SymbolRefAttr {
+    StringRef name = cast<SymbolRefAttr>(attr).getRootReference();
+    return (name == specName)
+               ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+               : cast<SymbolRefAttr>(attr);
+  };
+
+  SmallVector<Attribute> updatedMatchers;
+  SmallVector<Attribute> updatedActions;
+  for (auto matcherAttr : foreachMatchOp.getMatchers()) {
+    SymbolRefAttr updatedAttr = getUpdatedSymbol(matcherAttr);
+    updatedMatchers.push_back(updatedAttr);
+  }
+
+  for (auto actionAttr : foreachMatchOp.getActions()) {
+    SymbolRefAttr updatedAttr = getUpdatedSymbol(actionAttr);
+    updatedActions.push_back(updatedAttr);
+  }
+
+  // Apply the updated matchers and actions.
+  foreachMatchOp.setMatchersAttr(builder.getArrayAttr(updatedMatchers));
+  foreachMatchOp.setActionsAttr(builder.getArrayAttr(updatedActions));
+
+  op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+}
+
+// Retrieves the unique `ForeachMatchOp` inside the given `__kernel_config`
+// named sequence op. Asserts the op has the expected `__kernel_config` name and
+// `iree_codegen.tuning_spec_entrypoint` attribute.
+static ForeachMatchOp
+getForeachMatchOpFromKernelConfig(NamedSequenceOp namedSequenceOp) {
+  assert(namedSequenceOp.getSymName() == kKernelConfigSpecName &&
+         namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName) &&
+         "NamedSequenceOp must have the expected symbol name `__kernel_config` "
+         "and attribute `iree_codegen.tuning_spec_entrypoint`.");
+
+  auto foreachMatchOps = namedSequenceOp.getOps<ForeachMatchOp>();
+  assert(std::distance(foreachMatchOps.begin(), foreachMatchOps.end()) == 1 &&
+         "__kernel_config op should contain exactly one ForeachMatchOp.");
+
+  return *foreachMatchOps.begin();
```

Comment:
You can use `hasSingleElement` in the assert, or replace the thing with `getSingleElement` if our llvm is recent enough: https://github.com/llvm/llvm-project/pull/131508 . I see there was an integrate earlier today so this should be available when you rebase this PR.